core/src/main/scala/caliban/wrappers/FieldMetrics.scala

```diff
@@ -0,0 +1,90 @@
+package caliban.wrappers
+
+import caliban.CalibanError
+import caliban.ResponseValue
+import caliban.execution.FieldInfo
+import zio._
+import zio.metrics.Metric
+import zio.query.ZQuery
+import zio.metrics.MetricKeyType.Histogram
+import zio.metrics.MetricLabel
+
+object FieldMetrics {
+  private[caliban] val defaultBuckets = Histogram.Boundaries(
+    Chunk(
+      .005d, .01d, .025d, .05d, .075d, .1d, .25d, .5d, .75d, 1d, 2.5d, 5d, 7.5d, 10d
+    )
+  )
+
+  type Timing = (String, Long)
+  private def fieldDuration(name: String, field: String, buckets: Histogram.Boundaries) =
+    Metric.histogram(name, buckets).tagged("field", field)
+
+  private def fieldTotal(name: String, field: String) =
+    Metric.counter(name).tagged("field", field)
+
+  def wrapper(
+    totalLabel: String = "graphql_fields_total",
+    durationLabel: String = "graphql_fields_duration_seconds",
+    buckets: Histogram.Boundaries = defaultBuckets,
+    extraLabels: Set[MetricLabel] = Set.empty
+  ): Wrapper.EffectfulWrapper[Any] =
+    Wrapper.EffectfulWrapper(
+      for {
+        ref <- Ref.make(List.empty[Timing])
+      } yield fieldDuration(totalLabel, durationLabel, ref, buckets, extraLabels)
+    )
+
+  private def fieldDuration(
+    totalLabel: String,
+    durationLabel: String,
+    ref: Ref[List[Timing]],
+    buckets: Histogram.Boundaries,
+    extraLabels: Set[MetricLabel]
+  ): Wrapper.FieldWrapper[Any] =
+    new Wrapper.FieldWrapper[Any] {
+      override def wrap[R](
+        query: ZQuery[R, CalibanError.ExecutionError, ResponseValue],
+        info: FieldInfo
+      ): ZQuery[R, CalibanError.ExecutionError, ResponseValue] =
+        for {
+          summarized            <-
+            query
```

Comment:
Is that the part that `ZQuery` changes will make better?
Shall we wait for it? I see there is an open PR

Reply:
A maybe it's more about the `@@` on the data source. But same question anyway 😄

Reply:
No, I don't think so actually (but worth a try). The problem is that we're getting the execution time of all subsequent queries at the top level. Given `{ a (1s) { b (2s) { c (3s) } } }`, `a` will time in at 6 seconds. So what we do is keep track of the current path at each point and subtract the max execution time for our children, so e.g `c` will correctly get measured as `3`, `b` will get measured as `5 - 3 = 2` and `a` will be `6 - 5 = 1`. A bit messy / annoying but seems to work well!

The Apollo Tracing wrapper contains similar complexity due to the same reason.

Reply:
The PR was just merged so I dug into it a bit. It fixes the tracing (:tada:) but it doesn't really make a difference here. Since we need the `ResponseValue`, the [wrapper will effectively wrap all underlying queries needed to produce the `ResponseValue`](https://github.com/ghostdogpr/caliban/blob/1f2232c80d3df055986a324990bc9567be93ef04/core/src/main/scala/caliban/execution/Executor.scala#L139) and not just the immediate step needed to resolve the given field.

I guess what we could do is to introduce another kind of wrapper that only wraps the execution of a `QueryStep`, but I'm not sure if you could do much else with it rather than use it for metrics and tracing (since its input would essentially be a `ReducedStep[R]`). WDYT @ghostdogpr ?

Reply:
That's fine as it is 😄